sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java

```diff
@@ -801,40 +801,49 @@ private <T> Slice<T> sliceQuery(SqlQuerySpec querySpec,
         }
 
         final List<T> result = new ArrayList<>();
-        for (int index = 0; it.hasNext()
-            && index < pageable.getPageSize() + offsetForPageWithoutContToken; index++) {
+        while (result.size() < pageable.getPageSize() && it.hasNext()) {
+            for (int index = 0; it.hasNext()
+                && index < pageable.getPageSize() + offsetForPageWithoutContToken; index++) {
 
-            final JsonNode jsonNode = it.next();
-            if (jsonNode == null) {
-                continue;
-            }
+                final JsonNode jsonNode = it.next();
+                if (jsonNode == null) {
+                    continue;
+                }
 
-            if (index >= offsetForPageWithoutContToken) {
-                maybeEmitEvent(new AfterLoadEvent<>(jsonNode, returnType, containerName));
-                final T entity = mappingCosmosConverter.read(returnType, jsonNode);
-                result.add(entity);
+                if (index >= offsetForPageWithoutContToken) {
+                    maybeEmitEvent(new AfterLoadEvent<>(jsonNode, returnType, containerName));
+                    final T entity = mappingCosmosConverter.read(returnType, jsonNode);
+                    result.add(entity);
+                }
             }
-        }
-
-        final int contentSize = result.size();
 
-        int pageSize;
+            if (result.size() < pageable.getPageSize() && feedResponse.getContinuationToken() != null) {
+                feedResponseContentSize = pageable.getPageSize() - result.size();
+                feedResponseFlux = container
+                        .queryItems(querySpec, cosmosQueryRequestOptions, JsonNode.class)
+                        .byPage(feedResponse.getContinuationToken(), feedResponseContentSize);
 
-        if (contentSize < pageable.getPageSize()
-            && contentSize > 0) {
-            //  If the content size is less than page size,
-            //  this means, cosmosDB is returning less items than page size,
-            //  because of either RU limit, or payload limit
-
-            //  Set the page size to content size.
-            pageSize = contentSize;
-        } else {
-            pageSize = pageable.getPageSize();
+                feedResponse = feedResponseFlux
```

Comment:
Can we extract this logic to a function, since it is also being used above.

Reply:
Done.